[#205] Porting command line parsing infrastructure

lenhat509 opened the pull request:
Porting CLI infrastructure from pgagroal to pgmoneta

lenhat509 commented:
Hi @jesperpedersen @Jubilee101, I completed porting the command line parsing feature from pgagroal. Let me know if there is any concern.

jesperpedersen commented on src/libpgmoneta/utils.c:
```diff
@@ -3003,6 +3004,128 @@ pgmoneta_is_file_archive(char* file_path)
+   /* Iterate until find an accepted_arg_count that is equal or greater than the typed command arg_count */
+   arg_count = argc - offset;
+   int j;
+   for (j = 0; j < MISC_LENGTH; j++)
```
Define `int j` in the `for` loop

jesperpedersen replied:
Or move it to the top of the function

lenhat509 replied:
Got it. I'll move j to the top since it is used later.

jesperpedersen commented on src/libpgmoneta/utils.c:
```diff
@@ -3003,6 +3004,128 @@ pgmoneta_is_file_archive(char* file_path)
+              const struct pgmoneta_command command_table[],
+              size_t command_count)
+{
+#define EMPTY_STR(_s) (_s[0] == 0)
```
Hmm, isn't

```
if (s != NULL && strlen(s) > 0)
```

better ?

lenhat509 replied:
I think it makes the code a bit cleaner if we use macros. Subcommand is usually empty and we do the check frequently.

jesperpedersen commented on src/libpgmoneta/utils.c:
```diff
@@ -53,6 +53,7 @@
 #include <sys/time.h>
 #include <sys/stat.h>
 #include <sys/types.h>
+#include <err.h>
```
Sorting

lenhat509 replied:
I'll fix that. Thanks.

jesperpedersen commented:
You need to update `contrib/shell_comp`

jesperpedersen commented on src/libpgmoneta/utils.c:
```diff
@@ -3003,6 +3004,128 @@ pgmoneta_is_file_archive(char* file_path)
+              const struct pgmoneta_command command_table[],
+              size_t command_count)
+{
+#define EMPTY_STR(_s) (_s[0] == 0)
```
I think you should move this to `pgmoneta.h`

lenhat509 replied:
Just the macro, not the whole parse_command function, right?

jesperpedersen replied:
Just macro

lenhat509 commented:
> You need to update `contrib/shell_comp`

Sorry about missing that. I'll work on it.

lenhat509 commented:
Hi @jesperpedersen, I updated `contrib/shell_comp`, fixed a bug in `admin.c`, modified `cli.c` to handle failed commands nicely. In `cli.c`, it can display a helper function for the corresponding failed command instead of showing the entire menu. `usage()` will only be executed if the primary command does not exist. As for `admin.c`, the entire menu will be displayed since it is relatively small.

jesperpedersen commented:
Please, rebase

lenhat509 commented:
I added back version info for deprecated commands. Initially, I removed this info when I ported the code from pgagroal because I did not see version variables like `PGMONETA_MAJOR_VERSION` in pgmoneta. I just realized that I could define them in CMakeList.txt so I updated the code to show the version from which the commands were outdated (0.11).

jesperpedersen commented:
You need to update the documentation too...

Do a `git grep pgmoneta-cli` to see the files

lenhat509 commented:
Thanks for letting me know. I updated the doc. Let me know if I miss something else.

jesperpedersen commented:
Please, rebase

lenhat509 commented:
Hi @jesperpedersen , I have rebased it.

jesperpedersen commented:
Merged.

Thanks for your contribution !
